File: dotnet-blazor/Controllers/OrdersController.cs
using Microsoft.AspNetCore.Mvc;
using OrderMarginAnalysis.Models;
using OrderMarginAnalysis.Services;

namespace OrderMarginAnalysis.Controllers;

[ApiController]
[Route("api/[controller]")]
public class OrdersController : ControllerBase
{
    private readonly MockDataService _mockDataService;

    public OrdersController(MockDataService mockDataService)
    {
        _mockDataService = mockDataService;
    }

    /// <summary>
    /// Get all orders with optional filters
    /// </summary>
    /// <param name="searchQuery">Search in ID, customer name, or tracking number</param>
    /// <param name="dateFrom">Filter by order date from</param>
    /// <param name="dateTo">Filter by order date to</param>
    /// <param name="marketplace">Filter by marketplace</param>
    /// <param name="supplier">Filter by supplier</param>
    /// <param name="warehouse">Filter by warehouse</param>
    /// <param name="courier">Filter by courier</param>
    /// <param name="productCostStatus">Filter by product cost validation (OK/KO)</param>
    /// <param name="shippingCostStatus">Filter by shipping cost validation (OK/KO)</param>
    /// <returns>List of orders matching the filters</returns>
    [HttpGet]
    public ActionResult<List<Order>> GetOrders(
        [FromQuery] string? searchQuery = null,
        [FromQuery] DateTime? dateFrom = null,
        [FromQuery] DateTime? dateTo = null,
        [FromQuery] string? marketplace = null,
        [FromQuery] string? supplier = null,
        [FromQuery] string? warehouse = null,
        [FromQuery] string? courier = null,
        [FromQuery] string? productCostStatus = null,
        [FromQuery] string? shippingCostStatus = null)
    {
        var filters = new OrderFilters
        {
            SearchQuery = searchQuery,
            DateFrom = dateFrom,
            DateTo = dateTo,
            Marketplace = marketplace,
            Supplier = supplier,
            Warehouse = warehouse,
            Courier = courier,
            ProductCostStatus = productCostStatus,
            ShippingCostStatus = shippingCostStatus
        };

        var orders = _mockDataService.GetOrders(filters);
        return Ok(orders);
    }

    /// <summary>
    /// Get order by ID
    /// </summary>
    /// <param name="id">Order ID</param>
    /// <returns>Order with the specified ID</returns>
    [HttpGet("{id}")]
    public ActionResult<Order> GetOrderById(string id)
    {
        var order = _mockDataService.GetOrders()
            .FirstOrDefault(o => o.Id == id);

        if (order == null)
            return NotFound(new { message = $"Order {id} not found" });

        return Ok(order);
    }

    /// <summary>
    /// Get statistics for all orders or filtered orders
    /// </summary>
    /// <param name="searchQuery">Search in ID, customer name, or tracking number</param>
    /// <param name="dateFrom">Filter by order date from</param>
    /// <param name="dateTo">Filter by order date to</param>
    /// <param name="marketplace">Filter by marketplace</param>
    /// <param name="supplier">Filter by supplier</param>
    /// <param name="warehouse">Filter by warehouse</param>
    /// <param name="courier">Filter by courier</param>
    /// <param name="productCostStatus">Filter by product cost validation (OK/KO)</param>
    /// <param name="shippingCostStatus">Filter by shipping cost validation (OK/KO)</param>
    /// <returns>Statistics object</returns>
    [HttpGet("statistics")]
    public ActionResult<OrderStatistics> GetStatistics(
        [FromQuery] string? searchQuery = null,
        [FromQuery] DateTime? dateFrom = null,
        [FromQuery] DateTime? dateTo = null,
        [FromQuery] string? marketplace = null,
        [FromQuery] string? supplier = null,
        [FromQuery] string? warehouse = null,
        [FromQuery] string? courier = null,
        [FromQuery] string? productCostStatus = null,
        [FromQuery] string? shippingCostStatus = null)
    {
        var filters = new OrderFilters
        {
            SearchQuery = searchQuery,
            DateFrom = dateFrom,
            DateTo = dateTo,
            Marketplace = marketplace,
            Supplier = supplier,
            Warehouse = warehouse,
            Courier = courier,
            ProductCostStatus = productCostStatus,
            ShippingCostStatus = shippingCostStatus
        };

        var orders = _mockDataService.GetOrders(filters);
        var stats = _mockDataService.GetStatistics(orders);
        return Ok(stats);
    }

    /// <summary>
    /// Get list of all suppliers
    /// </summary>
    /// <returns>List of supplier names</returns>
    [HttpGet("suppliers")]
    public ActionResult<List<string>> GetSuppliers()
    {
        return Ok(_mockDataService.GetSuppliers());
    }

    /// <summary>
    /// Get list of all warehouses
    /// </summary>
    /// <returns>List of warehouse names</returns>
    [HttpGet("warehouses")]
    public ActionResult<List<string>> GetWarehouses()
    {
        return Ok(_mockDataService.GetWarehouses());
    }

    /// <summary>
    /// Get list of all couriers
    /// </summary>
    /// <returns>List of courier names</returns>
    [HttpGet("couriers")]
    public ActionResult<List<string>> GetCouriers()
    {
        return Ok(_mockDataService.GetCouriers());
    }

    /// <summary>
    /// Get list of all marketplaces
    /// </summary>
    /// <returns>List of marketplace names</returns>
    [HttpGet("marketplaces")]
    public ActionResult<List<string>> GetMarketplaces()
    {
        return Ok(_mockDataService.GetMarketplaces());
    }
}

File: vb-net/Controllers/OrdersController.vb
Imports System.Web.Mvc
Imports OrderMarginAnalysis.Models
Imports OrderMarginAnalysis.Services

Namespace Controllers
    ''' <summary>
    ''' Controller per la gestione degli ordini
    ''' </summary>
    Public Class OrdersController
        Inherits Controller

        Private ReadOnly _mockDataService As MockDataService

        Public Sub New()
            _mockDataService = New MockDataService()
        End Sub

        ' GET: Orders
        Public Function Index() As ActionResult
            Return View()
        End Function

        ' GET: api/orders - Restituisce ordini con filtri applicati
        Public Function GetOrders(searchQuery As String,
                                  dateFrom As String,
                                  dateTo As String,
                                  supplier As String,
                                  warehouse As String,
                                  courier As String,
                                  marketplace As String,
                                  productCostStatus As String,
                                  shippingCostStatus As String) As JsonResult

            Dim filters As New OrderFilters() With {
                .SearchQuery = searchQuery,
                .Supplier = supplier,
                .Warehouse = warehouse,
                .Courier = courier,
                .Marketplace = marketplace,
                .ProductCostStatus = productCostStatus,
                .ShippingCostStatus = shippingCostStatus
            }

            ' Parse date filters
            Dim parsedDateFrom As DateTime
            If Not String.IsNullOrWhiteSpace(dateFrom) AndAlso DateTime.TryParse(dateFrom, parsedDateFrom) Then
                filters.DateFrom = parsedDateFrom
            End If

            Dim parsedDateTo As DateTime
            If Not String.IsNullOrWhiteSpace(dateTo) AndAlso DateTime.TryParse(dateTo, parsedDateTo) Then
                filters.DateTo = parsedDateTo
            End If

            ' Ottieni ordini filtrati
            Dim orders = _mockDataService.GetOrders(filters)
            Dim statistics = _mockDataService.GetStatistics(filters)

            Dim result = New With {
                .orders = orders,
                .statistics = statistics
            }

            Return Json(result, JsonRequestBehavior.AllowGet)
        End Function

        ' GET: api/filter-options - Restituisce le opzioni per i filtri
        Public Function GetFilterOptions() As JsonResult
            Dim options = _mockDataService.GetFilterOptions()
            Return Json(options, JsonRequestBehavior.AllowGet)
        End Function

    End Class
End Namespace
